# 1. Analysis of Subtitles

In [1]:
subtitle_directory = 'MovieQA_benchmark/story/subtt'

In [2]:
# Get the names of files from subtitles_directory
from os import listdir
from os.path import isfile, join
def get_file_list(directory):
    subtitle_files = [f for f in listdir(directory) if isfile(join(directory, f))]
    return subtitle_files

In [3]:
subtitle_files = get_file_list(subtitle_directory)

In [4]:
len(subtitle_files)

408

In [5]:
subtitle_files[0:10]

['tt0034583.srt',
 'tt0036775.srt',
 'tt0056923.srt',
 'tt0061722.srt',
 'tt0068646.srt',
 'tt0071315.srt',
 'tt0071562.srt',
 'tt0074285.srt',
 'tt0075314.srt',
 'tt0076759.srt']

In [6]:
def get_subtitles(subtitle_file):
    with open(join(subtitle_directory, subtitle_files[i]), 'r') as f:
        read_data = f.read()
    return read_data.split('\n\n')

In [7]:
i = 0
subtitles = get_subtitles(subtitle_files[i])

In [8]:
subtitles[:10]

['1\n00:01:23,420 --> 00:01:26,218\nWith the coming\nof the Second World War...',
 '2\n00:01:26,557 --> 00:01:30,550\n...many eyes in imprisoned Europe\nturned hopefully, or desperately...',
 '3\n00:01:30,895 --> 00:01:33,557\n...toward the freedom of the Americas.',
 '4\n00:01:33,797 --> 00:01:36,425\nLisbon became\nthe great embarkation point.',
 '5\n00:01:36,634 --> 00:01:39,797\nBut not everybody could get\nto Lisbon directly.',
 '6\n00:01:40,237 --> 00:01:43,900\nAnd so a torturous, roundabout\nrefugee trail sprang up.',
 '7\n00:01:44,942 --> 00:01:46,569\nParis to Marseilles...',
 '8\n00:01:52,583 --> 00:01:55,245\n...across the Mediterranean to Oran.',
 '9\n00:02:02,593 --> 00:02:07,053\nThen, by train or auto or foot,\nacross the rim of Africa...',
 '10\n00:02:07,264 --> 00:02:09,755\n...to Casablanca in French Morocco.']

In [9]:
print(subtitles[0])

1
00:01:23,420 --> 00:01:26,218
With the coming
of the Second World War...


In [10]:
subtitles[0]

'1\n00:01:23,420 --> 00:01:26,218\nWith the coming\nof the Second World War...'

In [11]:
import re
def get_only_word_subtitle(subtitle):
    return re.sub(r'[0-9]+\n.+[0-9][0-9]:[0-9][0-9]:[0-9][0-9],[0-9][0-9][0-9]\n', r'', subtitle).replace('\n',' ') # remove time information

In [12]:
get_only_word_subtitle(subtitles[0])

'With the coming of the Second World War...'

In [13]:
def get_word_subtitles(subtitles):
    word_subtitles = list()
    for subtitle in subtitles:
        word_subtitles.append(get_only_word_subtitle(subtitle))
    return word_subtitles

In [14]:
word_subtitles = get_word_subtitles(subtitles)

In [15]:
print(len(subtitles))
print(len(word_subtitles))

1421
1421


In [16]:
import random
for i in range(100):
    print("[%d]: %s" % (i, random.choice(word_subtitles)))

[0]: Too bad. That's Casablanca's leading commodity.
[1]: I'm not fighting for anything anymore except myself.
[2]: And I hate this war so much.
[3]: The clipper to America.
[4]: A very puzzling fellow, this Rick. What sort is he?
[5]: Here they are.
[6]: I bet they're asleep in New York.
[7]: Victor, whatever I do, will you believe that I...?
[8]: Underground, like everywhere else.
[9]: She did her best to convince me that she was still in love with me.
[10]: Another precedent gone. This has been a very interesting evening.
[11]: If you will come--
[12]: Louis, I'll make a deal with you.
[13]: A couple of thousand less than I thought there would be.
[14]: Stay where you are. I don't know what you're talking about.
[15]: My dear, perhaps you have already observed that in Casablanca human life is cheap.
[16]: Since no one is to blame...
[17]: In many ways I am so much older than he is.
[18]: Sorry, monsieur, we were never to handle the police.
[19]: You won't dare to interfere with me h

In [17]:
string = 'qwe-A qt'

In [18]:
re.sub(r'\-([A-Z])', r' \1', string)

'qwe A qt'

In [19]:
# Remove '- ' that appears at the start.
def remove_dash(word_subtitles):
    for i in range(len(word_subtitles)):
        word_subtitles[i] = word_subtitles[i].replace('- ', '')
        word_subtitles[i] = word_subtitles[i].replace('--', ' ')
        word_subtitles[i] = re.sub(r'\-([A-Z])', r' \1', word_subtitles[i])
    return word_subtitles

In [20]:
word_subtitles = remove_dash(word_subtitles)

In [21]:
import random
for i in range(100):
    print("[%d]: %s" % (i, random.choice(word_subtitles)))

[0]: Excellent, captain.
[1]: My friends in the underground tell me that you have quite a record.
[2]: You must know it's very important I get out of Casablanca.
[3]: Besides, it isn't often that a man gets to display heroics before his wife.
[4]: ...because his insides have been kicked out.
[5]: A friend came and told me that Victor was alive.
[6]: ...I certainly won't give them to you now.
[7]: Rick, I think you know where those letters are.
[8]: Then we shall state another time and place. Tomorrow at 10 in the prefect's office.
[9]: I'll believe.
[10]: No, Rick. Not tonight. Especially tonight.
[11]: I'm sorry for asking.
[12]: I was in a German concentration camp for a year.
[13]: Practically no one.
[14]: Then I am going to the meeting Don't tell me where you're going.
[15]: ...and no one knows where those letters are.
[16]: The waters? What waters? We're in the desert.
[17]: Good night. Good night.
[18]: Morning. Signor Ferrari is the fat gent at the table.
[19]: And then Captain

In [22]:
# Concatenate if '...' appears in the start or end.
def concatenate_word_subtitles(word_subtitles):
    concat_word_subtitles = list()
    pattern = '...'
    for subtitle in word_subtitles:

        if subtitle[-len(pattern):] == pattern:
            # 'asdfasdf...'
            subtitle = subtitle[:-len(pattern)] # 'asdfasdf...' -> 'asdfasdf'

        if subtitle[:len(pattern)] == pattern:
            # '...asdfasdf'
            concat_word_subtitles[-1] = concat_word_subtitles[-1] + ' ' + subtitle[len(pattern):] 
        else:
            concat_word_subtitles.append(subtitle)
    return concat_word_subtitles

In [23]:
concat_word_subtitles = concatenate_word_subtitles(word_subtitles)

In [24]:
len(concat_word_subtitles) 

1345

In [25]:
len(word_subtitles)

1421

In [26]:
concat_word_subtitles[300:400]

["It's not particularly my beloved Paris.",
 'Can you imagine us in London?',
 'When you get there, ask me.',
 'Diplomatist.',
 'How about New York?',
 "Well, there are sections of New York, major, that I wouldn't advise you to try to invade.",
 "Who do you think will win the war? I haven't the slightest idea.",
 'Rick is completely neutral about everything. And that takes in the field of women too.',
 'You were not always so carefully neutral. We have a complete dossier on you.',
 '"Richard Blaine, American. Age: 37. Cannot return to his country."',
 'The reason is a little vague. We also know what you did in Paris.',
 "And also we know why you left Paris. Don't worry, we're not going to broadcast it.",
 'Are my eyes really brown?',
 'You will forgive my curiosity, Mr. Blaine. The point is an enemy of the Reich has come to Casablanca and we are checking up on anybody who can be of any help to us.',
 'My interest in whether Victor Laszlo stays or goes is purely a sporting one.',
 'In t

Well done!

In [27]:
s = 'This is a 0.9 stop. That is the sentence written by S. J. Cho!' 

In [28]:
re.sub(r'[a-z0-9]([\.?!])(\Z| )', r'[EOS]', s)

'This is a 0.9 sto[EOS]That is the sentence written by S. J. Ch[EOS]'

In [29]:
matchObj = re.search(r'[a-z0-9]([\.?!])(\Z| )', s)

In [30]:
matchObj.start(1)

18

In [31]:
matchObj.end(1)

19

In [32]:
s[matchObj.start(1):matchObj.end(1)]

'.'

In [33]:
matchObj.group(1)

'.'

In [34]:
def subtitle2sentences(subtitle):
    sentences = list()
    matchObj = re.search(r'[a-z0-9]([\.?!])(\Z| )', subtitle)
    while matchObj != None:
        sentences.append(s[:matchObj.end(1)])
        subtitle = subtitle[matchObj.end(1)+1:] # It is okay to be out of the range in terms of the array slicing.
        matchObj = re.search(r'[a-z0-9]([\.?!])(\Z| )', subtitle)
    return sentences

In [35]:
def subtitles2sentences(subtitles):
    sentences = list()
    for subtitle in concat_word_subtitles:
        sentences = sentences + subtitle2sentences(subtitle)
    return sentences

In [36]:
sentences = subtitles2sentences(concat_word_subtitles)

In [37]:
len(sentences)

1827

__The number of sentences in the first subtitle 'tt0034583.srt'__: 1827
* Uncompleted sentences are not counted.
  * Uncomplted sentence: a sentence that is not ended with '.', '!', or '?'.

# 2. For all subtitle files

In [38]:
subtitle_directory = 'MovieQA_benchmark/story/subtt'

In [39]:
# Get the names of files from subtitles_directory
from os import listdir
from os.path import isfile, join
def get_file_list(directory):
    subtitle_files = [f for f in listdir(directory) if isfile(join(directory, f))]
    return subtitle_files

In [40]:
subtitle_files = get_file_list(subtitle_directory)

In [41]:
def subtitle_files_to_sentences(subtitle_files):
    sentences = list()
    for subtitle_file in subtitle_files:
        subtitles = get_subtitles(subtitle_file)
        word_subtitles = get_word_subtitles(subtitles)
        word_subtitles = remove_dash(word_subtitles)
        concat_word_subtitles = concatenate_word_subtitles(word_subtitles)
        sentences = sentences + subtitles2sentences(concat_word_subtitles)
    return sentences

In [42]:
sentences = subtitle_files_to_sentences(subtitle_files)

In [43]:
len(sentences)

745416

__\# of sentences in all subtitles__: 745416

# 3. Codes without experiments

In [44]:
subtitle_directory = 'MovieQA_benchmark/story/subtt'

In [45]:
# Get the names of files from subtitles_directory
from os import listdir
from os.path import isfile, join
def get_file_list(directory):
    subtitle_files = [f for f in listdir(directory) if isfile(join(directory, f))]
    return subtitle_files

In [46]:
subtitle_files = get_file_list(subtitle_directory)

In [47]:
def get_subtitles(subtitle_file):
    with open(join(subtitle_directory, subtitle_files[i]), 'r') as f:
        read_data = f.read()
    return read_data.split('\n\n')

In [48]:
import re
def get_only_word_subtitle(subtitle):
    return re.sub(r'[0-9]+\n.+[0-9][0-9]:[0-9][0-9]:[0-9][0-9],[0-9][0-9][0-9]\n', r'', subtitle).replace('\n',' ') # remove time information

In [49]:
def get_word_subtitles(subtitles):
    word_subtitles = list()
    for subtitle in subtitles:
        word_subtitles.append(get_only_word_subtitle(subtitle))
    return word_subtitles

In [50]:
# Remove '- ' that appears at the start.
def remove_dash(word_subtitles):
    for i in range(len(word_subtitles)):
        word_subtitles[i] = word_subtitles[i].replace('- ', '')
        word_subtitles[i] = word_subtitles[i].replace('--', ' ')
        word_subtitles[i] = re.sub(r'\-([A-Z])', r' \1', word_subtitles[i])
    return word_subtitles

In [51]:
# Concatenate if '...' appears in the start or end.
def concatenate_word_subtitles(word_subtitles):
    concat_word_subtitles = list()
    pattern = '...'
    for subtitle in word_subtitles:

        if subtitle[-len(pattern):] == pattern:
            # 'asdfasdf...'
            subtitle = subtitle[:-len(pattern)] # 'asdfasdf...' -> 'asdfasdf'

        if subtitle[:len(pattern)] == pattern:
            # '...asdfasdf'
            concat_word_subtitles[-1] = concat_word_subtitles[-1] + ' ' + subtitle[len(pattern):] 
        else:
            concat_word_subtitles.append(subtitle)
    return concat_word_subtitles

In [52]:
def subtitle2sentences(subtitle):
    sentences = list()
    matchObj = re.search(r'[a-z0-9]([\.?!])(\Z| )', subtitle)
    while matchObj != None:
        sentences.append(s[:matchObj.end(1)])
        subtitle = subtitle[matchObj.end(1)+1:] # It is okay to be out of the range in terms of the array slicing.
        matchObj = re.search(r'[a-z0-9]([\.?!])(\Z| )', subtitle)
    return sentences

In [53]:
def subtitles2sentences(subtitles):
    sentences = list()
    for subtitle in concat_word_subtitles:
        sentences = sentences + subtitle2sentences(subtitle)
    return sentences

def subtitle_files_to_sentences(subtitle_files):
    sentences = list()
    for subtitle_file in subtitle_files:
        subtitles = get_subtitles(subtitle_file)
        word_subtitles = get_word_subtitles(subtitles)
        word_subtitles = remove_dash(word_subtitles)
        concat_word_subtitles = concatenate_word_subtitles(word_subtitles)
        sentences = sentences + subtitles2sentences(concat_word_subtitles)
    return sentences

In [54]:
sentences = subtitle_files_to_sentences(subtitle_files)

In [55]:
len(sentences)

745416

The following command takes very long time to be ended.

In [56]:
import re
words = list()
for sentence in sentences:
    words = words + re.split(r'\s', sentence) # r'\s' means a regular expression
    if len(words) % 100000 == 0:
        print(len(words))
print('Total number of words: %d' % (len(words)))

400000
900000
1200000
2300000
4500000
5700000
Total number of words: 5743416


# 4. Serialization: sentences, words

In [1]:
import cPickle

In [2]:
path_sentences = 'pickle_data/subtitle_sentences.p'
path_words = 'pickle_data/subtitle_words.p'

In [65]:
with open(path_sentences, 'wb') as f:
    cPickle.dump(sentences, f)

In [72]:
with open(path_words, 'wb') as f:
    cPickle.dump(words, f)

# 5. Loading the pickle files

In [1]:
import cPickle

In [2]:
path_sentences = 'pickle_data/subtitle_sentences.p'
path_words = 'pickle_data/subtitle_words.p'

In [3]:
with open(path_sentences, 'rb') as f:
    sentences = cPickle.load(f)

In [4]:
with open(path_words, 'rb') as f:
    words = cPickle.load(f)

In [5]:
len(sentences)

745416

In [6]:
len(words)

5743416

In [7]:
float(len(words)) / len(sentences)

7.704980842911877

__Average words per sentence__ : 7.704980842911877 = 7.7050